src/solvers/dg_p4est/containers.jl

```diff
@@ -0,0 +1,135 @@
+function init_elements(mesh::P4estMesh{NDIMS, RealT}, equations,
+                       basis, ::Type{uEltype}) where {NDIMS, RealT<:Real, uEltype<:Real}
+
+  nelements = ncells(mesh)
+  node_coordinates      = Array{RealT, NDIMS+2}(undef, NDIMS, ntuple(_ -> nnodes(basis), NDIMS)..., nelements)
+  jacobian_matrix       = Array{RealT, NDIMS+3}(undef, NDIMS, NDIMS, ntuple(_ -> nnodes(basis), NDIMS)..., nelements)
+  contravariant_vectors = similar(jacobian_matrix)
+  inverse_jacobian      = Array{RealT, NDIMS+1}(undef, ntuple(_ -> nnodes(basis), NDIMS)..., nelements)
+  surface_flux_values   = Array{uEltype, NDIMS+2}(undef, nvariables(equations),
+                                                  ntuple(_ -> nnodes(basis), NDIMS-1)..., NDIMS*2, nelements)
+
+  elements = ElementContainerP4est{NDIMS, RealT, uEltype, NDIMS+1, NDIMS+2, NDIMS+3}(
+      node_coordinates, jacobian_matrix, contravariant_vectors,
+      inverse_jacobian, surface_flux_values)
+
+  init_elements!(elements, mesh, basis)
+  return elements
+end
+
+@inline nelements(elements::ElementContainerP4est) = size(elements.node_coordinates, ndims(elements) + 2)
+@inline Base.ndims(::ElementContainerP4est{NDIMS}) where NDIMS = NDIMS
+
+Base.eltype(::ElementContainerP4est{NDIMS, RealT, uEltype}) where {NDIMS, RealT, uEltype} = uEltype
+
+
+struct InterfaceContainerP4est{NDIMS, uEltype<:Real, NDIMSP2} <: AbstractContainer
+  u::Array{uEltype, NDIMSP2}                  # [primary/secondary, variables, i, j, interface]
+  element_ids::Matrix{Int}                    # [primary/secondary, interfaces]
+  node_indices::Matrix{NTuple{NDIMS, Symbol}} # [primary/secondary, interfaces]
+end
+
+# Create interface container and initialize interface data in `elements`.
+function init_interfaces(mesh::P4estMesh, equations, basis,
+                         elements::ElementContainerP4est{NDIMS, RealT, uEltype}
+                         ) where {NDIMS, RealT<:Real, uEltype<:Real}
+  # Initialize container
+  n_interfaces = count_required_interfaces(mesh)
+
+  u = Array{uEltype, NDIMS+2}(undef, 2, nvariables(equations),
+                              ntuple(_ -> nnodes(basis), NDIMS-1)...,
+                              n_interfaces)
+  element_ids = Matrix{Int}(undef, 2, n_interfaces)
+  node_indices = Matrix{NTuple{NDIMS, Symbol}}(undef, 2, n_interfaces)
+
+  interfaces = InterfaceContainerP4est{NDIMS, uEltype, NDIMS+2}(u, element_ids, node_indices)
+
+  init_interfaces!(interfaces, mesh)
+
+  return interfaces
+end
+
+# Return number of interfaces
+@inline ninterfaces(interfaces::InterfaceContainerP4est) = size(interfaces.element_ids, 2)
+
+count_required_interfaces(mesh::P4estMesh) = 2 * ncells(mesh)
+
+
+# Convert Tuple node_indices to actual indices.
+# E.g., (:one, :i, :j) will be (1, i, j) for some i and j,
+# (:i, :end) will be (i, size[2]).
+function indexfunction(indices, size, dim, i, j=0)
+  if indices[dim] === :i
+    return i
+  elseif indices[dim] === :mi
```

Comment:
What does `mi` mean?

Reply:
I added a comment now. This means backward indexing for reversed orientations. `:mi` means something like "minus i".

Reply:
Why don't you write `:minus_i` then?

Reply:
You're right, `:mi` is not self-explanatory, but I'm not entirely happy with `:minus_i` either. This suggests that the index `-i` will be used.
Something like `:i_reversed` or `:i_backwards` would be self-explanatory, but I think that these don't sound good either. Do you have any other suggestions?

Reply:
I don't have a better suggestion right now. But these are both better than `:mi`, I think.